dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h

```diff
@@ -82,37 +68,37 @@ class ASTStencilBody : public ASTCodeGenCXX {
   std::string sparseArgName_ = "loc";
 
   bool parentIsReduction_ = false;
-  int parentReductionID_ = -1;
   bool parentIsForLoop_ = false;
-  bool firstPass_ = true;
   bool genAtlasCompatCode_ = false;
 
-  int currentBlock_ = -1;
-
-  std::map<int, std::stringstream> reductionMap_;
-  std::optional<MergeGroupMap> blockToMergeGroupMap_ = std::nullopt;
+  std::map<int, std::unique_ptr<ASTStencilBody>> reductionParser_;
+  int recursiveIterNest_;
 
   /// Nesting level of argument lists of stencil function *calls*
   int nestingOfStencilFunArgLists_;
 
   std::string makeIndexString(const std::shared_ptr<ast::FieldAccessExpr>& expr, std::string kiter);
   bool hasIrregularPentagons(const std::vector<ast::LocationType>& chain);
+  void evalNeighbourReduction(const std::shared_ptr<ast::ReductionOverNeighborExpr>& expr);
+  void generateNeighbourRedLoop(std::stringstream& ss) const;
+  std::string nbhLhsName(const std::shared_ptr<ast::Expr>& expr);
+  std::string pidxStr();
+  std::string nbhIterStr();
+  // symbol of the current neighbour index in a reduction loop
+  std::string nbhIdxStr();
+  // in case of nested reduction, this is the symbol of the index of parent reduction loop
+  std::string nbhIdxParentStr();
```

Comment:
I think those getters should be marked `const`

Reply:
done